### PR TITLE
[WIP] add nvidia riva

### PR DESCRIPTION
**Pull request by vipyne**
#### Please describe the changes in your PR. If it is addressing an issue, please reference that as well.

`$ python examples/foundational/07r-interruptible-riva-nim.py`

### CODE REVIEW

File: src/pipecat/services/riva.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    class InputParams(BaseModel):
+        language: Optional[Language] = Language.EN_US
+
+    def __init__(
+        self,
+        *,
+        api_key: str,
+        server: str = "grpc.nvcf.nvidia.com:443",
+        voice_id: str = "English-US.Female-1",
+        sample_rate_hz: int = 24000,
```
Comment (aconchillo) [inline src/pipecat/services/riva.py:49]: rename to `sample_rate` as other services.

File: src/pipecat/services/riva.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        # nvidia riva calls this 'function-id'
+        model: str = "0149dedb-2be8-4195-b9a0-e57e0e14f972",
+        params: InputParams = InputParams(),
+        quality: int = 20,
```
Comment (aconchillo) [inline src/pipecat/services/riva.py:53]: move `quality` to `InputParams` as:

```
quality: Optional[int] = 20
```

File: src/pipecat/services/riva.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        **kwargs,
+    ):
+        super().__init__(sample_rate=sample_rate_hz, **kwargs)
+        self._api_key = api_key
+        self._voice_id = voice_id
+        self._sample_rate_hz = sample_rate_hz
+        self._language_code = params.language
+        self.quality = quality
```
Comment (aconchillo) [inline src/pipecat/services/riva.py:61]: `self._quality = params.quality`

File: src/pipecat/services/riva.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        self.set_model_name("fastpitch-hifigan-tts")
+        self.set_voice(voice_id)
+
+        metadata = [
+            ["function-id", model],
+            ["authorization", f"Bearer {api_key}"],
+        ]
+        auth = riva.client.Auth(None, True, server, metadata)
+
+        self.service = riva.client.SpeechSynthesisService(auth)
```
Comment (aconchillo) [inline src/pipecat/services/riva.py:72]: self._service`

File: src/pipecat/services/riva.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+    async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        def read_audio_responses():
+            try:
+                responses = self.service.synthesize_online(
```
Comment (aconchillo) [inline src/pipecat/services/riva.py:77]: `self._service`

File: src/pipecat/services/riva.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    class InputParams(BaseModel):
+        language: Optional[Language] = Language.EN_US
+
+    def __init__(
+        self,
+        *,
+        api_key: str,
+        server: str = "grpc.nvcf.nvidia.com:443",
+        # nvidia calls this 'function-id'
+        model: str = "1598d209-5e27-4d3c-8079-4751568b1081",
```
Comment (aconchillo) [inline src/pipecat/services/riva.py:122]: what is this UUID? is it specific to our account?

Comment (vipyne) [reply src/pipecat/services/riva.py:122]: It is not specific to our account.  It is the parakeet / riva ASR function id
https://build.nvidia.com/nvidia/parakeet-ctc-1_1b-asr/api
We could just remove "model" and call this function_id.

File: src/pipecat/services/riva.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        model: str = "0149dedb-2be8-4195-b9a0-e57e0e14f972",
+        params: InputParams = InputParams(),
+        **kwargs,
+    ):
+        super().__init__(sample_rate=sample_rate, **kwargs)
+        self._api_key = api_key
+        self._voice_id = voice_id
+        self._sample_rate = sample_rate
+        self._language_code = params.language
+        self.quality = params.quality
```
Comment (aconchillo) [inline src/pipecat/services/riva.py:61]: `self._quality = params.quality`

File: src/pipecat/services/riva.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+    async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        def read_audio_responses():
+            try:
+                responses = self._service.synthesize_online(
+                    text,
+                    self._voice_id,
+                    self._language_code,
+                    sample_rate_hz=self._sample_rate,
+                    audio_prompt_file=None,
+                    quality=self.quality,
```
Comment (aconchillo) [inline src/pipecat/services/riva.py:83]: `self._quality`

File: src/pipecat/services/riva.py (hunk 6 above)
Comment (aconchillo) [reply src/pipecat/services/riva.py:122]: Oh, then yes, I think we should call this `function_id`.

File: src/pipecat/services/riva.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+        self.set_model_name("parakeet-ctc-1.1b-asr")
+
+        metadata = [
+            ["function-id", model],
+            ["authorization", f"Bearer {api_key}"],
+        ]
+        auth = riva.client.Auth(None, True, server, metadata)
+
+        self.asr_service = riva.client.ASRService(auth)
```
Comment (aconchillo) [inline src/pipecat/services/riva.py:151]: self._asr_service = riva...

File: src/pipecat/services/riva.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+        await self._stop_tasks()
+
+    async def _stop_tasks(self):
+        self._thread_task.cancel()
+        await self._thread_task
+        self._response_task.cancel()
+        await self._response_task
+
+    def _response_handler(self):
+        responses = self.asr_service.streaming_response_generator(
```
Comment (aconchillo) [inline src/pipecat/services/riva.py:208]: `self._asr_service`

File: src/pipecat/services/riva.py [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+
+    def __init__(
+        self,
+        *,
+        api_key: str,
+        server: str = "grpc.nvcf.nvidia.com:443",
+        voice_id: str = "English-US.Female-1",
+        sample_rate: int = 24000,
+        # nvidia riva calls this 'function-id'
+        model: str = "0149dedb-2be8-4195-b9a0-e57e0e14f972",
```
Comment (aconchillo) [inline src/pipecat/services/riva.py:52]: Same here, `function_id`.

**Review by aconchillo — APPROVED** [review]

**Comment by aconchillo** [comment]
LGTM!

**Pull request merged**